tests/unit/test_lib_per_unit_requires.py

```diff
@@ -199,3 +199,49 @@ def test_ipu_events(self):
         # AND a revoked event fires
         after = self.harness.charm.revoked_event_count
         self.assertGreater(after, before)
+
+
+class TestInterlibDependency(unittest.TestCase):
```

Comment:
@rbarry82 could you please take a look at this test?
Is that what you had in mind on the [catalogue PR](https://github.com/canonical/catalogue-k8s-operator/pull/4#issuecomment-1326823357)?

Reply:
Not exactly -- to make it reflective of what I had in mind (or what traefik-route is already doing), `ingress_per_unit` would need changes.

Put differently, if you [walk the code back](https://github.com/canonical/traefik-k8s-operator/blob/main/lib/charms/traefik_k8s/v1/ingress_per_unit.py#L840), `_get_urls_from_relation_data` still will not be called (nor will it update `.url`) until an event comes through. And event which would trigger the observer _after_ it's resolved, because `copy(x)` still doesn't create a closure anyway IIRC.

Reply:
> `ingress_per_unit` would need changes.

Yes of course. This is just a failing test - I haven't started touching the code.
I meant: does this test, as written, capture the use case you had in mind, for which I'd then need to change the IPU code so the test passes.

Reply:
Ah, in that case, mostly yes. I guess that when I think of "inter-lib dependency", I would still think of it as simply:

```python
self.dependee = { "value": self.ipu.url }
```

Without the `copy()`, but it's not super important

Reply:
The copy is intentional, to convey that we do not rely on passing object reference.